eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala

```diff
@@ -0,0 +1,74 @@
+package fr.acinq.eclair.io
+
+import akka.actor.typed.delivery.DurableProducerQueue.TimestampMillis
+import akka.actor.typed.eventstream.EventStream
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.{ActorRef, Behavior}
+import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.eclair.NodeParams
+import fr.acinq.eclair.channel.ChannelCreated
+import fr.acinq.eclair.io.Monitoring.Metrics
+import fr.acinq.eclair.io.Peer.Disconnect
+
+/**
+ * A singleton actor that limits the total number of incoming connections from peers that do not have channels with us.
+ *
+ * When a new incoming connection request is received, the Switchboard should send an
+ * [[IncomingConnectionsTracker.TrackIncomingConnection]] message.
+ *
+ * When the number of tracked peers exceeds `eclair.peer-connection.max-no-channels`, send [[Peer.Disconnect]] to
+ * the tracked peer with the oldest incoming connection.
+ *
+ * When a tracked peer disconnects or adds a channel, we will stop tracking that peer.
+ *
+ * We do not need to track peers that disconnect because they will terminate if they have no channels.
+ * Likewise, peers with channels will disconnect and terminate when their last channel closes.
+ *
+ * Note: Peers on the sync whitelist are not tracked.
+*/
+object IncomingConnectionsTracker {
+  // @formatter:off
+  sealed trait Command
+
+  case class TrackIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private case class ForgetIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private[io] case class CountIncomingConnections(replyTo: ActorRef[Int]) extends Command
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, switchboard: ActorRef[Disconnect]): Behavior[Command] = {
+    Behaviors.setup { context =>
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[PeerDisconnected](c => ForgetIncomingConnection(c.nodeId)))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelCreated](c => ForgetIncomingConnection(c.remoteNodeId)))
+      new IncomingConnectionsTracker(nodeParams, switchboard).tracking(Map())
+    }
+  }
+}
+
+private class IncomingConnectionsTracker(nodeParams: NodeParams, switchboard: ActorRef[Disconnect]) {
+  import IncomingConnectionsTracker._
+
+  private def tracking(incomingConnections: Map[PublicKey, TimestampMillis]): Behavior[Command] = {
```

Comment:
It's probably worth a comment on this actor explaining that this doesn't completely solve DoS issues: an attacker could trivially keep opening new connection with us in a loop with random `node_id`s, and that will force us to constantly disconnect every one else (because honest peers won't even have enough time to open a channel before they get disconnected).

This is somewhat ok, because it only affects non-whitelisted incoming connections from peers with whom we didn't have previous channels, so it only limits your ability to receive new incoming liquidity "for free". To work around this, node operators should probably add many known popular nodes to their `sync-whitelist` to make sure they avoid the rate-limits.

I don't think it's this actor's responsibility to do more than what it's already doing: the only way to meaningfully protect against an attacker constantly opening connections with random `node_id`s would be to rate-limit based on IP/AS buckets, which we don't want to do at the application level. This is something that should be done at the TCP level before connections reach eclair, by adding a "standard" DoS-protecting TCP load balancer or something similar.

Reply:
We could do like we do in the `PendingChannelsRateLimiter` and keep two lists: one for public and one for private nodes so that only free liquidity from private nodes could be DoS'd that way. On the other hand, it seems like searching our list of public nodes for each incoming connection could also be a form of resource DoS. I'll just update the comment and we can iterate if an attack materializes.

Reply:
Is this too wordy? [3ac3d49](https://github.com/ACINQ/eclair/pull/2601/commits/3ac3d49a416b43d39eca31c8fde49b3650bd1beb)